### ISSUE TITLE
[java] Incorrect ImmutableField error on field with both declaration and constructor initializer

### ISSUE THREAD

**Issue by stevenschlansker** [pmd/pmd#4046]
**Affects PMD Version:**
6.47.0

**Rule:**

Please provide the rule name and a link to the rule documentation:
[Design.ImmutableField](https://pmd.github.io/pmd/pmd_rules_java_design.html#immutablefield)

**Description:**

ImmutableField sees an assignment that happens in a declaration _or_ constructor, which would indicate the field may be final. But in the case where it is assigned in the declaration _and_ constructor, it is no longer true that it may be final, but the rule still triggers.

**Code Sample demonstrating the issue:**

```
public class NotSoImmutableField {
    private int foo = 2;

    NotSoImmutableField() {
    }

    NotSoImmutableField(final int foo) {
        this.foo = foo;
    }

    public int getFoo() {
        return foo;
    }
}
```

**Expected outcome:**

The code is correct. PMD reports a violation:

```
[INFO] --- maven-pmd-plugin:3.17.0:check (basepom.default) @ wsj-util ---
[INFO] PMD version: 6.47.0
[INFO] PMD Failure: NotSoImmutableField:4 Rule:ImmutableField Priority:3 Private field 'foo' could be made final; it is only initialized in the declaration or constructor..
```

This warning is not correct. `foo` may not be final, since it is initialized both at declaration and in one of the constructors. Following the PMD recommendation results in a (correct) compiler error.

**Running PMD through:**
Maven

**Comment by jsotuyod** [comment]
This is no longer reproducible under PMD 7.0.0

Having said that, I don't agree this was a FP, and i'd argue it not being reported in 7.0.0 is a regression / false negative.

The sample code **can** effectively be rewritten to make `foo` immutable, as it's never written after initialization, for instance, the code could be rewritten as:

```java
public class NotSoImmutableField {
    private final int foo;

    NotSoImmutableField() {
        this(2);
    }

    NotSoImmutableField(final int foo) {
        this.foo = foo;
    }

    public int getFoo() {
        return foo;
    }
}
```

I'm creating a new issue to track this as a FN.
